Remove dependency on internal micro-bunzip in favor of dynamically linking against libbz2
--- a/snesreader/Makefile
+++ b/snesreader/Makefile
@@ -6,7 +6,7 @@
 c     := $(compiler) -std=gnu99
 cpp   := $(subst cc,++,$(compiler)) -std=gnu++0x
 flags := -O3 -I. -Iobj -fomit-frame-pointer
-link  :=
+link  := -lbz2
 
 ifeq ($(platform),x)
   flags := -fPIC $(flags)
@@ -26,7 +26,7 @@
 # unrar
 objects += archive arcread coder crc encname extract getbits model rarvm rarvmtbl rawread suballoc unicode unpack unpack15 unpack20 unrar unrar_misc unrar_open
 # micro-bunzip
-objects += micro-bunzip
+#objects += micro-bunzip
 # jma
 objects += jma jcrc32 lzmadecode 7zlzma iiostrm inbyte lzma winout
 
--- a/snesreader/snesreader.cpp
+++ b/snesreader/snesreader.cpp
@@ -8,10 +8,10 @@
 
 #include "fex/fex.h"
 #include "libjma/jma.h"
-extern "C" char* uncompressStream(int, int);  //micro-bunzip
 
 #define QT_CORE_LIB
 #include <QtGui>
+#include <bzlib.h>
 
 #include <nall/file.hpp>
 #include <nall/string.hpp>
@@ -147,29 +147,63 @@
   return false;
 }
 
+// Uncompresses bzip2 compressed file filename into data and sets size to sizeof
+// (uncompressed) data
 bool snesreader_load_bz2(const char *filename, uint8_t *&data, unsigned &size) {
-  //TODO: need a way to get the size of a bzip2 file, so we can pre-allocate
-  //a buffer to decompress into memory. for now, use a temporary file.
+  // This function is not from upstream. It was rewritten to use the dynamically
+  // linked libbz2 rather than the included micro-bunzip
 
-  string name = "/tmp/.bz2_temporary_decompression_object";
-  FILE *wr;
-  wr = fopen_utf8(name, "wb");
-  if(!wr) {
-    //try the local directory
-    name = ".bz2_temporary_decompression_object";
-    wr = fopen_utf8(name, "wb");
-    //can't get write access, so give up
-    if(!wr) return false;
-  }
-
-  FILE *fp = fopen_utf8(filename, "rb");
-  uncompressStream(fileno(fp), fileno(wr));
-  fclose(fp);
-  fclose(wr);
-
-  bool success = snesreader_load_normal(name, data, size);
-  unlink(name);
-  return success;
+  FILE * f;
+  BZFILE * b;
+  int nBuf;
+  const int bufsize = 8192;
+  char buf[bufsize];
+  int bzerror;
+  char * long_buf = NULL;
+
+  // Open file
+  if ((f = fopen(filename, "rb")) == NULL)
+    return false;
+  b = BZ2_bzReadOpen(&bzerror, f, 0, 0, NULL, 0);
+  if (bzerror != BZ_OK) {
+    BZ2_bzReadClose (&bzerror, b);
+    fclose(f);
+    return false;
+  }
+
+  // Read file into buf bufsize chunks at a time
+  size = 0;
+  while (bzerror == BZ_OK) {
+    nBuf = BZ2_bzRead(&bzerror, b, buf, bufsize);
+    if (bzerror == BZ_OK || bzerror == BZ_STREAM_END) {
+      // Extend long_buf with data from buf
+      size += nBuf;
+      long_buf = (char *) realloc(long_buf, size);
+      memcpy(long_buf + size - nBuf, buf, nBuf);
+    }
+  }
+  // File is fully loaded. Make sure error status is as expected.
+  if (bzerror != BZ_STREAM_END) {
+    BZ2_bzReadClose(&bzerror, b);
+    fclose(f);
+    free(long_buf);
+    return false;
+  } else {
+    BZ2_bzReadClose(&bzerror, b);
+    fclose(f);
+  }
+
+  // Copy temporary buffer long_buf into data
+  data = new uint8_t[size];
+  memcpy(data, long_buf, size);
+  free(long_buf);
+
+/*  // For debugging
+  f = fopen("/tmp/bsnes.rom", "wb");
+  fwrite(data, 1, size, f);
+  fclose(f); */
+
+  return true;
 }
 
 bool snesreader_load_jma(const char *filename, uint8_t *&data, unsigned &size) {
